bluepysnap/nodes/node_population.py

```diff
@@ -94,28 +95,102 @@ def _node_sets(self):
         return self._circuit.node_sets
 
     @cached_property
-    def _data(self):
-        """Collect data for the node population as a pandas.DataFrame."""
-        nodes = self._population
-        categoricals = nodes.enumeration_names
-
-        _all = nodes.select_all()
-        result = pd.DataFrame(index=pd.RangeIndex(_all.flat_size, name="node_ids"))
-
-        for attr in sorted(nodes.attribute_names):
-            if attr in categoricals:
-                enumeration = np.asarray(nodes.get_enumeration(attr, _all))
-                values = np.asarray(nodes.enumeration_values(attr))
-                # if the size of `values` is large enough compared to `enumeration`, not using
-                # categorical reduces the memory usage.
-                if values.shape[0] < 0.5 * enumeration.shape[0]:
-                    result[attr] = pd.Categorical.from_codes(enumeration, categories=values)
-                else:
-                    result[attr] = values[enumeration]
-            else:
-                result[attr] = nodes.get_attribute(attr, _all)
-        for attr in sorted(utils.add_dynamic_prefix(nodes.dynamics_attribute_names)):
-            result[attr] = nodes.get_dynamics_attribute(attr.split(DYNAMICS_PREFIX)[1], _all)
+    def _cache(self):
+        """Cached DataFrame of nodes, to be accessed through _get_data()."""
+        return pd.DataFrame(index=pd.RangeIndex(self.size, name="node_ids"))
+
+    def _get_libsonata_selection(self, node_ids):
+        """Return a libsonata Selection from the given node_ids."""
+        if node_ids is None:
+            return libsonata.Selection([(0, self.size)])
```

Comment:
Using directly `libsonata.Selection([(0, self.size)])` is 100x faster because `self._population` isn't cached, and it's still 10x faster even when `self._population` is cached:

```
In [1]: import bluepysnap
   ...: import libsonata
   ...: 
   ...: path = "/gpfs/bbp.cscs.ch/project/proj30/tickets/BLPY-289_improve_get/circuit/circuit_config.json"
   ...: c = bluepysnap.Circuit(path)
   ...: nodes = c.nodes["root__neurons"]
   ...: 
   ...: print(nodes.size)
   ...: _population = nodes._population
   ...: 
   ...: %timeit libsonata.Selection([(0, nodes.size)])
   ...: %timeit _population.select_all()
   ...: %timeit nodes._population.select_all()
   ...: 
74124987
1.25 µs ± 13.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
14.5 µs ± 22 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
136 µs ± 81.7 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
However, the times are under 1 ms, so the performance shouldn't be an issue, unless some loop is involved (as in the tests in https://github.com/BlueBrain/snap/pull/211)
Do you still prefer using `self._population.select_all()` in this case?

Reply:
ok, good to know; keep the current one then :)